### ISSUE TITLE
High Level functions

### ISSUE THREAD

**Issue by AnkurDedania** [Azure/azure-storage-python#110]
I was looking at https://azure.microsoft.com/en-us/documentation/articles/storage-python-how-to-use-blob-storage/#list-the-blobs-in-a-container and was wondering if the azure storage sdk can provide a similar functionality.

Something close too the following example?
```
class AzureListBlobsIterator(collections.Iterable):

    def __init__(self, service, container_name, prefix=None, **kwargs):
        self.service = service
        self.container_name = container_name
        self.prefix = prefix
        self.kwargs = kwargs

    def __iter__(self):
        marker = None
        while True:
            batch = self.service.list_blobs(self.container_name, self.prefix, marker=marker, **self.kwargs)
            
            yield from batch
            if not batch.next_marker:
                break
            marker = batch.next_marker

class AzureListPrefixesIterator(collections.Iterable):

    def __init__(self, service, container_name, prefix=None, **kwargs):
        self.service = service
        self.container_name = container_name
        self.prefix = prefix
        self.kwargs = kwargs

    def __iter__(self):
        marker = None
        while True:
            batch = self.service.list_blobs(self.container_name, self.prefix, marker=marker, **self.kwargs)
            
            yield from batch.prefixes
            if not batch.next_marker:
                break
            marker = batch.next_marker
```

**Comment by emgerner-msft** [comment]
This will be addressed in our upcoming release per issue #57. In fact, the code is already in the dev branch if you'd like to take a look.

**Comment by jehine-MSFT** [comment]
Hello @AbsoluteMSTR ,

We have added lazy enumeration for all our listing APIs in the latest 0.30 release (#132). Each API will now return a generator which automatically follows the continuation tokens when the next item in the list is needed. Please review our design and let us know what you think of it. We'd love to hear your feedback!

Thanks for the suggestion!

**Comment by AbsoluteMSTR** [comment]
@jehine-MSFT @emgerner-msft

I like the implementation, For `ListGenerator` https://github.com/Azure/azure-storage-python/blob/master/azure/storage/models.py#L38. Shouldn't you define it as `collection.Iterable` instead of just `object`

**Comment by emgerner-msft** [comment]
Why? Actually curious, what is the gain here?

We generally followed the python documentation such as [this](https://wiki.python.org/moin/Generators) and existing implementations such as that for the python management libraries [here](https://github.com/Azure/autorest/blob/python/ClientRuntimes/Python/msrest/msrest/paging.py). We didn't see anywhere folks extending that.

**Comment by AbsoluteMSTR** [comment]
It helps with libraries that verify an instance of a data, example usage I found in pandas.

https://github.com/pydata/pandas/blob/f5550e2bb702bf0133eebfcc3cdce4dcd0cf735b/pandas/io/html.py#L694

**Comment by emgerner-msft** [comment]
Ah okay, just type checking. No real functional difference which is probably why we didn't see this in any examples we looked at. Looking at the [base class definition](https://docs.python.org/3/library/collections.abc.html#collections-abstract-base-classes) I don't see a problem with extending this, though it does look like there's [a change](http://pydocs2cn.readthedocs.org/en/latest/library/collections.html#abcs-abstract-base-classes) since Python 2. As we support both, we'll have to be careful with the imports.

@zooba if he sees any caveats
@annatisch since she works on the management libs I referenced

**Comment by zooba** [comment]
Seems fine to me.

**Comment by emgerner-msft** [comment]
Alright, I'll add this to our next release then. You should see it in the dev branch soon-ish if you're in a hurry to grab source. Thanks for the feedback!

**Comment by emgerner-msft** [comment]
This should be released as of version 0.31.0. Thanks again for the feedback!
